drivers/adc/ad738x/iio_ad738x.c

```diff
@@ -0,0 +1,200 @@
+/***************************************************************************//**
+* @file   iio_ad738x.c
+* @brief  Implementation of AD738X IIO interface
+********************************************************************************
+* Copyright (c) 2024 Analog Devices, Inc.
+* Copyright (c) 2024 BayLibre, SAS.
+* All rights reserved.
+*
+* This software is proprietary to Analog Devices, Inc. and its licensors.
+* By using this software you agree to the terms of the associated
+* Analog Devices Software License Agreement.
+*******************************************************************************/
+#ifdef IIO_SUPPORT
+
+#include <stdio.h>
+#include <math.h>
+
+#include "ad738x.h"
+#include "iio_ad738x.h"
+#include "iio.h"
+#include "no_os_util.h"
+#include "no_os_alloc.h"
+#include "no_os_error.h"
+
+#define AD738X_CH(_idx) {\
+	.name = "voltage" #_idx, \
+	.ch_type = IIO_VOLTAGE,\
+	.ch_out = IIO_DIRECTION_INPUT,\
+	.indexed = true,\
+	.channel = _idx,\
+	.scan_index = _idx,\
+	.scan_type = &ad738x_iio_scan_type,\
+	.attributes = ad738x_iio_ch_attributes,\
+}
+
+#define CHN_STORAGE_BITS	16
+#define ADC_RESOLUTION		16
+#define ADC_REF_VOLTAGE		3300
+#define ADC_MAX_COUNT_UNIPOLAR  (uint32_t)((1 << ADC_RESOLUTION) - 1)
+
+struct ad738x_iio {
+	struct ad738x_dev *drv_dev;
+	struct iio_device *iio_dev;
+};
+
+/**
+ * @brief Read the debug register value
+ * @param device - Pointer to IIO device instance
+ * @param reg - Register address to read from
+ * @param val - Pointer to variable to read data into
+ * @return 0 in case of success, negative value otherwise
+ */
+static int32_t iio_ad738x_debug_reg_read(void *device, uint32_t reg,
+		uint32_t *val)
+{
+	struct ad738x_iio_dev *iio_dev = device;
+	struct ad738x_dev *dev = iio_dev->ad738x_dev;
+
+	return ad738x_spi_reg_read(dev,  (uint8_t) reg, (uint16_t *)val);
```

Comment:
I'm pretty sure this may create bugs on processor with a certain endianess. When you cast that pointer to uint16_t * you're essentially saying the ad738x_spi_reg_read to fill in the uppermost or lowermost (endianess based) 2 bytes of the 4 bytes that is being used to represent val. And when val is interpreted as uint32_t, in one of the scenarios you'll end up with a wrong value. You might need to create an temporary variable here to achieve what you want.